src/librustc/diagnostics.rs

```diff
@@ -1431,6 +1431,21 @@ fn make_child<'elve>(x: &mut &'elve isize, y: &mut &'elve isize) {
 ```
 "##,
 
+E0317: r##"
+An `if` expression without an `else` block is required to have the type `()`.
+This error occurs when the `if` block has a type other than `()`. For example:
+
+```compile_fail,E0317
+fn main() {
+    let x = 5;
+    let a = if x == 5 { 1 };
```

Comment:
This makes sense to me -- even if you added an else, given that the expected type of this expression is `()`, the body would have to be changed. That is, the body and the fn signature are in conflict. We choose to blame the body. You could make a case the other way, but blaming the *if* seems wrong, since there is nothing you can change about the `if` itself (without changing something else) that would make this code compile.

Not that the compiler thinks this hard when selecting what to blame; it's semi-random. Just saying that in this case the *span* makes sense. It'd be nice if we could give a better error message though, explaining where the `()` came from.

I think that Elm does some nice tracking here. We should copy them. =)